File: csharp/6.ZigzagConversion.cs
namespace Leetcode.ZigzagConversion;

using System.Text;

public class Solution {
    public string Convert(string s, int numRows) {
        if (numRows == 1)
            return s;

        var rows = Enumerable.Range(0, numRows)
            .Select(_ => new StringBuilder())
            .ToArray();

        int currentRow = 0;
        bool goingDown = false;

        foreach (var c in s) {
            rows[currentRow].Append(c);

            if (currentRow == 0 || currentRow == numRows - 1)
                goingDown = !goingDown;

            currentRow += goingDown ? 1 : -1;
        }

        return String.Join("", rows.Select(row => row.ToString()));
    }
}

File: typescript/6.ZigzagConversion.ts
function convert(s: string, numRows: number): string {
    if (numRows === 1) return s

    const rows = [] as string[][]

	let row = 0
	let goingDown = false
	for (let i = 0; i < s.length; i++) {
		rows[row] ??= []
		rows[row].push(s[i])

		if (row === 0 || row === numRows - 1)
            goingDown = !goingDown

		row += goingDown ? 1 : -1
	}

	return rows.reduce((prev, curr) => [...prev, curr.join('')], []).join('')
}
